File: Forth_Interpreter/Division.cpp
#include "Division.hpp"
#include "exceptions.hpp"

#include <memory>

bool Division::Run () {
    if (GeneralDataForExecutors.Stack->empty ()) {
        throw Exception_EmptyStack();
    }

    int a = GeneralDataForExecutors.Stack->top ();
    GeneralDataForExecutors.Stack->pop ();

    if (GeneralDataForExecutors.Stack->empty ()) {
        throw Exception_EmptyStack();
    }

    int b = GeneralDataForExecutors.Stack->top ();
    GeneralDataForExecutors.Stack->pop ();

    if (a == 0) {
        throw Exception_DivisionByZero();
    }

    GeneralDataForExecutors.Stack->push(b / a);

    return false;
}
 
namespace ForthFactoryRegistrations {
    FactoryRegistrationCommands<Division> Command_division ("/");
}

File: Forth_Interpreter_On_Java/src/main/java/ru/nsu/yurakov/Forth/Division/Division.java
package ru.nsu.yurakov.Forth.Division;

import java.util.EmptyStackException;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ru.nsu.yurakov.Forth.ForthCommands.ForthCommands;
import ru.nsu.yurakov.Forth.ForthInterpreter.ForthInterpreter;
import ru.nsu.yurakov.Forth.ForthRuntimeException.ForthRuntimeException;

/**
 * Division the first number from the stack from the second<br>
 * Pushes the result on the stack
 */
public class Division extends ForthCommands{
    
    @Override
    public boolean Run() throws ForthRuntimeException {
        LOGGER.info("Was launched " + this.getClass().getName());
        int a = 0;
        int b = 0;
        try {
            a = GeneralDataForExecutors.stack.peek();
            GeneralDataForExecutors.stack.pop();
        } catch (RuntimeException ex) {
            throw new ForthRuntimeException("Stack is empty");
        }
        try {
            b = GeneralDataForExecutors.stack.peek ();
            GeneralDataForExecutors.stack.pop ();
        } catch (RuntimeException ex) {
            GeneralDataForExecutors.stack.push(a);
            throw new ForthRuntimeException("Stack is empty");
        }

        if (a == 0) {
            throw new ForthRuntimeException("Division by zero");
        }
        else {
            GeneralDataForExecutors.stack.push(b / a);
        }
    
        return false;
    }

    private static final Logger LOGGER = LogManager.getLogger(Division.class);
}
